patroni/utils.py

```diff
@@ -911,3 +912,31 @@ def enable_keepalive(sock: socket.socket, timeout: int, idle: int, cnt: Optional
 
     for opt in keepalive_socket_options(timeout, idle, cnt):
         sock.setsockopt(*opt)
+
+
+def shell_quote(opt: str) -> str:
+    """Quote a string intended to be used as part of a shell command.
+
+    Applies quoting on POSIX compliant systems only, on ``win32`` platforms
+    will just return the original string unquoted.
+
+    :param opt: String to be quoted
+    :returns: Quoted string
+    """
+    if sys.platform != 'win32' and not is_single_quoted_string(opt):
+        return quote(opt)
+    return opt
+
+
+def is_single_quoted_string(string: str) -> bool:
+    """Check if a string is a single fully quoted string and not multi-quoted using shlex.
+
+    :param string: The input string to check.
+    :returns: ``True`` if the input string is a single fully quoted string, ``False`` otherwise.
+    """
+    lexer = shlex(string)
+    # Next 2 assignments ensure shlex does not split the string on whitespace
+    lexer.whitespace_split = False
+    lexer.whitespace = ''
+    tokens = list(lexer)
+    return len(tokens) == 1 and tokens[0][0] == tokens[0][-1] and tokens[0][0] in ('"', "'")
```

Comment:
Unfortunately it doesn't work with one of my examples:
```python
In [1]: from shlex import quote, shlex

In [2]: a = 'asd \' ads'

In [3]: c = quote(a)

In [4]: lexer = shlex(c)

In [5]: lexer.whitespace_split = False

In [6]: lexer.whitespace = ''

In [7]: tokens = list(lexer)

In [8]: len(tokens)
Out[8]: 3  # <-- HERE
```

Reply:
> Unfortunately it doesn't work with one of my examples:
...
> In [8]: len(tokens)
> Out[8]: 3  # <-- HERE
> ```

Hmnn, drat.

Reply:
> Maybe we need a behave test for this.

I am not so sure. As far as I know TDE isn't available as an extension and installing it on GH is an additional effort.
IMO it would be enough just to check that both cases work fine (simplified, you need to add missing initdb arguments):
```python
subprocess.call(['initdb', '--key-wrap-command="openssl enc -e -pbkdf2 -out %p -pass pass:trustno1"'])
subprocess.call(["initdb", "--key-wrap-command='openssl enc -e -pbkdf2 -out %p -pass pass:trustno1'"])
```

Reply:
`process_user_options()` already checks that the value is `str`, hence we always pass the `str` object to the `shell_quote()`.

Reply:
> > Maybe we need a behave test for this.
> 
> I am not so sure. As far as I know TDE isn't available as an extension and installing it on GH is an additional effort. IMO it would be enough just to check that both cases work fine (simplified, you need to add missing initdb arguments):
> 
> ```python
> subprocess.call(['initdb', '--key-wrap-command="openssl enc -e -pbkdf2 -out %p -pass pass:trustno1"'])
> subprocess.call(["initdb", "--key-wrap-command='openssl enc -e -pbkdf2 -out %p -pass pass:trustno1'"])
> ```

So, I ran this test to mimic what the existing bootstrap code is doing
```
#!/usr/bin/env python3

import sys
import subprocess

args = ['--encoding=UTF8',
        '--data-checksums',
        '--auth-local=peer',
        '--auth-host=scram-sha-256',
        '--data-encryption',
        '--key-wrap-command="openssl enc -e -pbkdf2 -out %p -pass pass:trustno1"',
        '--key-unwrap-command="openssl enc -d -pbkdf2 -in %p -pass pass:trustno1"']

options = ['-o', ' '.join(args)]

res = subprocess.call(['/usr/lib/edb-as/15/bin/pg_ctl', 'initdb', '-D', sys.argv[1]] + options)
```

I also inverted the quoting, double enclosing single. Both worked and resulted in a single quoted value in the postgresql.conf.
```
data_encryption_key_unwrap_command = 'openssl enc -d -pbkdf2 -in %p -pass pass:trustno1'
```

Note that a direct call to initdb does not work, as you gave above in your example to test, the quotes are interpreted literally. It results in an error
```
setting up data encryption ... sh: 1: openssl enc -e -pbkdf2 -out d2/pg_encryption/key.bin -pass pass:trustno1: not found
initdb: error: could not run command "'openssl enc -e -pbkdf2 -out d2/pg_encryption/key.bin -pass pass:trustno1'": command not found
```

So, this only works because the arguments are passed as one via the options var which combines all args as a single argument passed to pg_ctl. If this was handled directly as a call to initdb the args should be sent without quotes but separate items in the list of args.

```
#!/usr/bin/env python3

import sys
import subprocess

args = ['--encoding=UTF8', 
        '--data-checksums', 
        '--auth-local=peer', 
        '--auth-host=scram-sha-256', 
        '--data-encryption', 
        '--key-wrap-command=openssl enc -e -pbkdf2 -out %p -pass pass:trustno1', 
        '--key-unwrap-command=openssl enc -d -pbkdf2 -in %p -pass pass:trustno1']

res = subprocess.call(['/usr/lib/edb-as/15/bin/initdb'] + args + [sys.argv[1]])
```

Reply:
Yeah, it seems that just calling initdb directly would simplify a lot of things.
I don't already recall why it was done like this. Probably because we just can...

Reply:
I don't propose to change that now for this PR, although I would be happy to add that as something to look into if you think it's worthwhile?

Reply:
Calling initdb directly  would eliminate the need to have shell_quote() and requires_quoting() all together.

Reply:
I want to say - lets do it the right way.
Quoting-unquoting-checking seems to created many strange problems that better solved by avoiding them.

P.S., process_user_options() docstring is beautiful, we should keep it!

Reply:
OK, I'll take a look at how to do that. It would reduce clutter.

Reply:
Of course we have the possible case now where a user is relying on the current way this works, i.e. double quoting.

```
initdb:
  - key-wrap-command: "'openssl enc -e -pbkdf2 -out %p -pass pass:trustno1'"
```

If we want to be backwards compatible we'd need to `unquote` that string. Then we're back where we were before!

Reply:
Oh, you already have it somewhere...

> we'd need to unquote that string

It could be as simple as:
```python
def unquote(value: str) -> str:
    try:
        ret = split(value)
        ret = ret[0] if len(ret) == 1 else value
    except Exception:
        ret = value
    return ret
```

Reply:
OK pushed the code for this. I noticed that the options parser method used in bootstrap is also used by pg_basebackup. I think it should be OK though as it doesn't go through pg_ctl as initdb did.

Reply:
Hmm, I don't see any new commits...

Reply:
Gak, sorry pushed to our internal repo. Should be there now.